src/hub.js

```diff
@@ -1237,203 +1203,129 @@ document.addEventListener("DOMContentLoaded", async () => {
   APP.messageDispatch = messageDispatch;
   document.getElementById("avatar-rig").messageDispatch = messageDispatch;
 
-  let isInitialJoin = true;
+  hubChannel.channel = hubPhxChannel;
+  hubChannel.presence = new Presence(hubPhxChannel);
+  const { rawOnJoin, rawOnLeave } = denoisePresence(presenceEventsForHub(events, hubId));
```

Comment:
Phoenix calls the `onJoin` and `onLeave` handlers whenever we receive any changes for the `meta`(s) associated with any `presence` key. (We use `session_id`'s as our keys.)

Here we introduce `denoisePresence` to make it easier to tell which of these things is happening:
- we receive a meta for a NEW presence (aka `session_id`), 
- ALL metas for a given presence (`session_id`) have left, or
- the meta for a given presence (`session_id`) changes.

This allows us to (for example) write `events.on("hub:join", ({key, meta})=>{ //... })` to handle a new participant joining the hub channel.